### ISSUE TITLE
how to load a model with Scatter operation in C++ 

### ISSUE THREAD

**Issue by bs1119** [rusty1s/pytorch_scatter#212]
I am trying to load a Pytorch model in C++ with pytorch_scatter operation --> A.pt 

I first build C++ API of pytorch_scatter. 
And, then in vs 2017, I include in additional include directores;

    pytorch_scatter-2.0.5\csrc
    pytorch_scatter-2.0.5\csrc\cpu

also, in linker --> additional Library dictories:

    \pytorch_scatter-2.0.5\build3\Release
 
The project build with no errors. 
But,  the executable still can not load the model A.pt. 

However, if I modify and save the model without using scatter operation --> A_no_Scatter.pt
The executable can load the model. 

Here is my C++ file:

    #include <torch/script.h> // One-stop header.
    #include <iostream>
    #include <memory>
    #include <torch/torch.h>
    #include <scatter.h>
    
    int main(int argc, const char* argv[]) {
	    if (argc != 2) {
		    std::cerr << "usage: example-app <path-to-exported-script-module>\n";
		    return -1;
	    }
    
	    std::cout << argv[1] << "\n";
    
    
	    torch::jit::script::Module module;
	    try {
		    // Deserialize the ScriptModule from a file using torch::jit::load().
		    module = torch::jit::load(argv[1]);
	    }
	    catch (const c10::Error& e) {
		    std::cerr << "error loading the model\n";
		    std::cout << e.what() << std::endl;
		    return -1;
	    }
    
	    std::cout << "ok\n";
    }

Can anyone tell me what else I need to do？

Thanks.

**Comment by rusty1s** [comment]
You need to import `<torchscatter/scatter.h>`, see https://github.com/rusty1s/pytorch_geometric/issues/1718#issuecomment-779780517.

**Comment by bs1119** [comment]
> You need to import `<torchscatter/scatter.h>`, see [rusty1s/pytorch_geometric#1718 (comment)](https://github.com/rusty1s/pytorch_geometric/issues/1718#issuecomment-779780517).

Thanks a lot. 
I saw your example. But, since I am building torch-scatter in VS2017. 
I only end up with a Release folder containing:  torchscatter.dll, torchscatter.lib, and torchscatter.exp. 

How should I prepare the cmakelist.txt? 

currently, I have:
  
    cmake_minimum_required(VERSION 3.12 FATAL_ERROR)
    project(example-app)
    
    find_package(Torch REQUIRED)
    
    
    if(NOT Torch_FOUND)
        message(FATAL_ERROR "Pytorch Not Found!")
    endif(NOT Torch_FOUND)
    
    add_library(torchscater SHARED IMPORTED)
    set_property(TARGET torchscater PROPERTY IMPORTED_LOCATION c:/tmp/torchscatter.dll)
    set_property(TARGET torchscater PROPERTY IMPORTED_IMPLIB c:/tmp/torchscatter.lib)
    
    add_executable(example-app DeepLearning.cpp)
    target_link_libraries(example-app ${TORCH_LIBRARIES})
    target_link_libraries(example-app -Wl,--no-as-needed torchscater)
    set_property(TARGET example-app PROPERTY CXX_STANDARD 11)

**Comment by bs1119** [comment]
> You need to import `<torchscatter/scatter.h>`, see [rusty1s/pytorch_geometric#1718 (comment)](https://github.com/rusty1s/pytorch_geometric/issues/1718#issuecomment-779780

Maybe my question is:
How to do “make install"  TorchScatter C++ API in visual studio 2017？

After build in vs2017, I didn;t see a install folder of torchscatter in anywhere. 

Sorry. It's my first time to work in a windows + visual studio 2017 environment. 
And, I have to load my model in windows environment.

**Comment by bs1119** [comment]
> You need to import `<torchscatter/scatter.h>`, see [rusty1s/pytorch_geometric#1718 (comment)](https://github.com/rusty1s/pytorch_geometric/issues/1718#issuecomment-779780517).

Hi, I figured out and was able to install TorchScatter. 
I also include #include <torchscatter/scatter.h>

But still I can not load my model. 

I similplied my model to just scatter_max, as below: 
    
    class Classifier_sactter_jit(torch.nn.Module):
        def __init__(self):
            super(Classifier_sactter_jit, self).__init__()
            
        def forward(self, x1, batch ): 
            
            c1 = scatter_max(x1, batch, dim=0, dim_size=int(batch.max().item() + 1))
            return(c1)


Here is my C++ code: 
    
    #include <torch/script.h> // One-stop header.
    #include <iostream>
    #include <memory>
    #include <torch/torch.h>
    #include <torchscatter/scatter.h>
    
    int main(int argc, const char* argv[]) {
	    if (argc != 2) {
		    std::cerr << "usage: example-app <path-to-exported-script-module>\n";
		    return -1;
	    }
    
	    
	    torch::DeviceType device_type;
	    if (torch::cuda::is_available()) {
		    std::cout << "CUDA available! Training on GPU." << std::endl;
		    device_type = torch::kCUDA;
	    }
	    else {
		    std::cout << "Training on CPU." << std::endl;
		    device_type = torch::kCPU;
	    }
	    torch::Device device(device_type);
    
	    std::cout << argv[1] << "\n";
    
    
	    torch::jit::script::Module module;
	    try {
		    // Deserialize the ScriptModule from a file using torch::jit::load().
		    module = torch::jit::load(argv[1]);
	    }
	    catch (const c10::Error& e) {
		    std::cerr << "error loading the model\n";
		    std::cout << e.what() << std::endl;
		    return -1;
	    }
    
	    std::cout << "ok\n";
    }

**Comment by rusty1s** [comment]
What is the error message?

**Comment by bs1119** [comment]
> What is the error message?

There is no error message, since I only have release mode. But, I suspect it is related:

[https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html]

"There is one crucial detail embedded in the above example: The -Wl,--no-as-needed prefix to the warp_perspective link line. This is required because we will not actually be calling any function from the warp_perspective shared library in our application code. We only need the TORCH_LIBRARY function to run. Inconveniently, this confuses the linker and makes it think it can just skip linking against the library altogether. On Linux, the -Wl,--no-as-needed flag forces the link to happen (NB: this flag is specific to Linux!). **There are other workarounds for this. The simplest is to define some function in the operator library that you need to call from the main application. This could be as simple as a function void init(); declared in some header, which is then defined as void init() { } in the operator library. Calling this init() function in the main application will give the linker the impression that this is a library worth linking against.** Unfortunately, this is outside of our control, and we would rather let you know the reason and the simple workaround for this than handing you some opaque macro to plop in your code. "

Can you provide an C++ example on calling functions of torchscatter, similar as for torchsparse:

thanks.  

    #include <iostream>
    #include <torch/torch.h>
    #include <torchsparse/sparse.h>
    
    int main() {
      torch::Tensor tensor = torch::tensor({0, 0, 0, 0, 1, 1, 1});
      std::cout << tensor << std::endl;
      std::cout << ind2ptr(tensor, 2) << std::endl;
    }

**Comment by rusty1s** [comment]
That would look something like (untested):
```
#include <iostream>
#include <torch/torch.h>
#include <torchscatter/scatter.h>

int main() {
  torch::Tensor src = torch::tensor({0.5, 0.4, 0.1, 0.6});
  torch::Tensor index = torch::tensor({0, 0, 1, 1});
  std::cout << src << std::endl;
  std::cout << index << std::endl;
  std::cout << scatter_sum(src, index, 0) << std::endl;
}
```

**Comment by bs1119** [comment]
> That would look something like (untested):
> 
> ```
> #include <iostream>
> #include <torch/torch.h>
> #include <torchscatter/scatter.h>
> 
> int main() {
>   torch::Tensor src = torch::tensor({0.5, 0.4, 0.1, 0.6});
>   torch::Tensor index = torch::tensor({0, 0, 1, 1});
>   std::cout << src << std::endl;
>   std::cout << index << std::endl;
>   std::cout << scatter_sum(src, index, 0) << std::endl;
> }
> ```

Thanks for the quick reply, but now there is a build error: 
   
             Error C2660	'scatter_sum': function does not take 3 arguments	


Did I install the wrong version? Here is the definition of scatter_sum:

    torch::Tensor scatter_sum(torch::Tensor src, torch::Tensor index, int64_t dim,
                              torch::optional<torch::Tensor> optional_out,
                              torch::optional<int64_t> dim_size);

**Comment by rusty1s** [comment]
Sorry, that was my fault:
```
#include <iostream>
#include <torch/torch.h>
#include <torchscatter/scatter.h>

int main() {
  torch::Tensor src = torch::tensor({0.5, 0.4, 0.1, 0.6});
  torch::Tensor index = torch::tensor({0, 0, 1, 1});
  std::cout << src << std::endl;
  std::cout << index << std::endl;
  std::cout << scatter_sum(src, index, 0, torch::nullopt, torch::nullopt) << std::endl;
}
```

**Comment by DBraun** [comment]
Why is the directory` #include <torchscatter/scatter.h>` instead of `#include <pytorch_scatter/csrc/scatter.h>`?

**Comment by rusty1s** [comment]
This is directly taken from https://github.com/pytorch/vision/tree/main/examples/cpp/hello_world.

**Comment by DBraun** [comment]
But in this repository (pytorch_scatter), the directory "torchscatter" doesn't exist, so it's confusing me. Does this directory structure exist somewhere else? For instance, if I use pytorch_scatter as a submodule in my project, I will end up not using torchscatter/scatter.h. Instead I would use something closer to pytorch_scatter/csrc/scatter.h

**Comment by rusty1s** [comment]
I see. I have never tried to add `pytorch_scatter` as a submodule, but I'm happy to look into this. Do you have a small example to reproduce?
